### PR TITLE
IGNITE-13101 Fixed uncompleted futures leak on node stop in distributed metastorage.

### PR DESCRIPTION
**Pull request by ibessonov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java [hunk 1]
```diff
@@ -1119,7 +1199,16 @@ private void onAckMessage(
...
-        GridFutureAdapter<Boolean> fut = updateFuts.remove(msg.requestId());
+        GridFutureAdapter<Boolean> fut;
+
+        updateFutsStopLock.readLock().lock();
+
+        try {
+            fut = updateFuts.remove(msg.requestId());
+        }
+        finally {
+            updateFutsStopLock.readLock().unlock();
```
Comment (AMashenkov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java:1210]: Readlock looks useless here.
Anyway, you'll finish the future in few lines below.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java [hunk 2]
```diff
@@ -1070,6 +1125,31 @@ else if (!isClient && ver.id() > 0) {
         return fut;
     }
 
+    /** */
+    private GridFutureAdapter<Boolean> validateBeforeWrite(String key) throws IgniteCheckedException {
+        boolean supported;
+
+        try {
+            supported = isSupported(ctx);
```
Comment (AMashenkov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java:1133]: Is it really necessary to validate all the nodes on every operation?
Do we allow old nodes to connect to new grid? If no, then may be we can skip this check if once found all the nodes support the feature?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java [hunk 3]
```diff
@@ -1033,14 +1058,29 @@ else if (!isClient && ver.id() > 0) {
...
 
         GridFutureAdapter<Boolean> fut = new GridFutureAdapter<>();
 
-        updateFuts.put(reqId, fut);
+        updateFutsStopLock.readLock().lock();
+
+        try {
+            if (stopped) {
+                fut.onDone(nodeStoppingException());
+
```
Comment (AMashenkov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java:1075]: Does it make sense to move this into 'validateBeforeWrite' ?
'validateBeforeWrite' can be executed under read-lock as in normal case it always return null.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java (hunk 2 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java:1142]: Every component does this, changing the approach is out of the scope. And yes, older node can enter the cluster sometimes.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java (hunk 1 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java:1219]: You're right, did that automatically.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java (hunk 3 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java:1075]: Yes it does, thank you!

**Pull request closed without merging**